cmd/internal/database/meilisearch/meilisearch.go

```diff
@@ -0,0 +1,329 @@
+	if err != nil {
+		return err
+	}
+	meilisearchVersion, err := db.getBinaryVersion(ctx)
+	if err != nil {
+		return err
+	}
+	if (dbVersion.Major() == meilisearchVersion.Major()) && (dbVersion.Minor() == meilisearchVersion.Minor()) {
+		db.log.Infow("no version difference, no upgrade required", "database-version", dbVersion, "binary-version", meilisearchVersion)
+		return nil
+	}
+	if dbVersion.GreaterThan(meilisearchVersion) {
+		db.log.Errorw("database is newer than meilisearch binary, aborting", "database-version", dbVersion, "binary-version", meilisearchVersion)
+		return fmt.Errorf("database is newer than meilisearch binary")
+	}
+	if _, err := os.Stat(db.latestStableDumpDst); errors.Is(err, fs.ErrNotExist) {
+		return fmt.Errorf("%q is not present, no upgrade possible, maybe no backup was running before", db.latestStableDumpDst)
+	}
+
+	db.log.Infow("start upgrade", "from", dbVersion, "to", meilisearchVersion)
+
+	err = os.Rename(db.dbdir, db.dbdir+".old")
+	if err != nil {
+		return fmt.Errorf("unable to rename dbdir: %w", err)
+	}
+
+	err = db.importDump(ctx, db.latestStableDumpDst)
+	if err != nil {
+		return err
+	}
+	db.log.Infow("upgrade done and new data in place", "took", time.Since(start))
+	return nil
+}
+
+func (db *Meilisearch) importDump(ctx context.Context, dump string) error {
+	var (
+		err  error
+		cmd  *exec.Cmd
+		g, _ = errgroup.WithContext(ctx)
+	)
+
+	g.Go(func() error {
+		args := []string{"--import-dump", dump, "--master-key", db.apikey}
+		db.log.Infow("execute meilisearch", "args", args)
+
+		cmd = exec.CommandContext(ctx, meilisearchCmd, args...) // nolint:gosec
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		err = cmd.Run()
+		if err != nil {
+			return fmt.Errorf("unable import dump %w", err)
+		}
+		db.log.Info("import of dump finished")
+		return nil
+	})
+
+	// TODO big databases might take longer, not sure if 100 attempts are enough
+	// must check how long it take max with backoff ?
+	err = retry.Do(func() error {
+		v, err := db.client.Version()
+		if err != nil {
+			return err
+		}
+		healthy := db.client.IsHealthy()
+		if !healthy {
+			return fmt.Errorf("meilisearch does not report healthiness")
+		}
+		db.log.Infow("meilisearch started after importing the dump, killing it", "version", v)
+		return cmd.Process.Signal(syscall.SIGTERM)
+	}, retry.Attempts(100), retry.Context(ctx))
+	if err != nil {
+		return err
+	}
+	err = g.Wait()
+	if err != nil {
+		// sending a TERM signal will always result in a error response.
+		db.log.Infow("importing dump terminated but reported an error which can be ignored", "error", err)
+	}
+	return nil
+}
+
+// moveDumpToBackupDir move all dumps to the backupdir
+// also create a stable last stable dump for later upgrades
+func (db *Meilisearch) moveDumpToBackupDir(ctx context.Context) error {
+	dumps, err := filepath.Glob(db.dumpdir + "/*.dump")
+	if err != nil {
+		return fmt.Errorf("unable to find dumps %w", err)
+	}
```

Comment:
According to the doc `Glob` can return `nil` for matches, which is not an error. The error should be returned and also the length of matches should not be zero length.

Reply:
Also done